### ISSUE TITLE
virt.running: serial_type, console parameters missing

### ISSUE THREAD

**Issue by ewenmcneill** [saltstack/salt#53081]
### Description of Issue/Question

[`salt.modules.virt.init`](https://github.com/saltstack/salt/blob/develop/salt/modules/virt.py#L1217) has parameters to define a `libvirt` serial console:

*    [`serial_type: pty`](https://github.com/saltstack/salt/blob/develop/salt/modules/virt.py#L1381)

*   [`console: True`](https://github.com/saltstack/salt/blob/develop/salt/modules/virt.py#L1384)

which I would like to use, as I define all my virtual machines with serial console (on a libvirt / KVM / qemu server that is a `salt` minion).

Unfortunately [`salt.state.virt.running`](https://github.com/saltstack/salt/blob/develop/salt/states/virt.py#L239) (and other related functions) provide no way to pass `serial_type` or `console` through to `salt.modules.virt.init`.

Ideally there would either (a) be `serial_type` / `console` parameters on the `salt.states.virt.running` function, or (b) some way to pass `**kwargs` style additional initialisation parameters from the `salt.state.virt.running` definition through to the `salt.modules.virt.init` module so that these additional parameters could be provided.

### Setup

```
debian_unstable:
    virt.running:
      - cpu: 1
      - mem: 512
      - vm_type: kvm
      - serial_type: pty
      - console: True
      - disk_profile: None
      - disks:
          - name: vda
            source_file: /dev/ssd/debian_unstable_vda
            format: raw
            model: virtio
      - install: False
      - nic_profile: None
      - interfaces:
          - name: eth0
            type: bridge
            source: virbr0
            mac: 00:01:02:03:04:05
            model: virtio
```

### Steps to Reproduce Issue

Try to `state.highstate` the libvirt minion with a definition like the one above.  It fails with:

```
naosr620.naos.co.nz:
    ERROR executing 'state.highstate': 'serial_type' and 'console' are invalid keyword arguments for 'virt.running'
```

and inspecting the code (see links above) there is no obvious way to pass the `serial_type` or `console` values in via a `virt` state.

### Versions Report

`salt` master is currently on Debian Jessie (8), with Salt 2019.2.0 on Python 2.7; `salt` minions are all Salt 2019.2.0, and in the case of the minion where I ran into this issue the `salt` minion is on Ubuntu 18.04 LTS with Salt 20.19.2.0 on Python 2.7.

```
ewen@noc:~$ salt --versions-report
Salt Version:
           Salt: 2019.2.0
 
Dependency Versions:
           cffi: 0.8.6
       cherrypy: Not Installed
       dateutil: 2.2
      docker-py: Not Installed
          gitdb: 0.5.4
      gitpython: 0.3.2 RC1
          ioflo: Not Installed
         Jinja2: 2.9.4
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.2
   mysql-python: Not Installed
      pycparser: 2.10
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.9 (default, Sep 25 2018, 23:32:58)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 14.4.0
           RAET: Not Installed
          smmap: 0.8.2
        timelib: Not Installed
        Tornado: 4.4.3
            ZMQ: 4.0.5
 
System Versions:
           dist: debian 8.11 
         locale: ANSI_X3.4-1968
        machine: i686
        release: 4.9.0-0.bpo.9-686-pae
         system: Linux
        version: debian 8.11 
 
ewen@noc:~$ 
```

**Comment by ewenmcneill** [comment]
In case anyone else runs into this, I've currently hacked around the inability to specify the `serial_type` / `console` via the `states.virt` by editing `salt.states.virt` on the *minion* in question to just always pass the values I want into `salt.modules.virt.init`, with this diff (2019.2):

```
ewen@naosr620:~$ diff -u /usr/lib/python2.7/dist-packages/salt/states/virt.py.old /usr/lib/python2.7/dist-packages/salt/states/virt.py
--- /usr/lib/python2.7/dist-packages/salt/states/virt.py.old	2019-02-16 19:13:46.000000000 +1300
+++ /usr/lib/python2.7/dist-packages/salt/states/virt.py	2019-05-17 16:14:04.272620242 +1200
@@ -381,6 +381,12 @@
 
     '''
 
+    # Edited 2019-05-17 to add serial_type, console values, hard coded
+    # as these cannot be provided via the state at present
+    #
+    serial_type = 'pty'
+    console     = True
+
     ret = {'name': name,
            'changes': {},
            'result': True,
@@ -457,7 +463,9 @@
                                   priv_key=priv_key,
                                   connection=connection,
                                   username=username,
-                                  password=password)
+                                  password=password,
+                                  serial_type=serial_type,
+                                  console=console)
             ret['changes'][name] = 'Domain defined and started'
             ret['comment'] = 'Domain {0} defined and started'.format(name)
     except libvirt.libvirtError as err:
ewen@naosr620:~$ 
```

It seems to work, giving:

```
root@naosr620:~# grep -A 10 serial /etc/libvirt/qemu/debian_unstable.xml | grep -B 99 '</console'
    <serial type='pty'>
      <target type='isa-serial' port='0'>
        <model name='isa-serial'/>
      </target>
    </serial>
    <console type='pty'>
      <target type='serial' port='0'/>
    </console>
root@naosr620:~# 
```

injected into the virtual domain definition, which means that `console VMNAME` will work in the `libvirt` `virsh` interface (and I was able to log in as expected, which is how I got the details about [`sshd` not starting due to randomness delays](https://github.com/saltstack/salt/issues/53087#issuecomment-493314169)).

Arguably those values should be the *default* values (ie, serial console, optionally with graphical UI as well).

Ewen

**Comment by garethgreenaway** [comment]
@ewenmcneill Thanks for the report. Having those additional parameters available would be a great additional.  I'm marking this as a feature request and if you would be interested in submitting a PR with the above changes that would be greatly appreciated.

**Comment by ewenmcneill** [comment]
I'm busy for a while, but yes if no one else gets to it before me, I'll figure out how to get the relevant parameters from the config YAML through to the `salt.virt.modules.init` call and create a PR.  It's likely to be a few weeks at least before I have time.  I'm happy if someone else wants to do it first :-)

Ewen

**Comment by ewenmcneill** [comment]
Commenting to keep issue open; yes I'd still like the ability to specify serial console parameters.   (And it's still on my "to do when I have time" list if no one else does it first.)

Ewen

**Comment by cbosdo** [comment]
This has been implemented in 3003. Could you check that the added features match your requirements? Hopefully it should be flexible enough to cover this issue.

**Comment by ewenmcneill** [comment]
Thanks for the heads up @cbosdo!

For anyone else playing along at home, the actually applied change seems to be https://github.com/saltstack/salt/issues/58844 (issue) and https://github.com/saltstack/salt/pull/56430 (PR; both found via [3003 release notes](https://docs.saltproject.io/en/latest/topics/releases/3003.html)).

From reading the change diff (but not yet testing it), I *think*:

    serials:
        - type: pty
    consoles:
        - type: pty

will result in the same output I'm relying on now in the libvirt file, ie generating something like:

```
    <serial type='pty'>
      <target type='isa-serial' port='0'>
        <model name='isa-serial'/>
      </target>
    </serial>
    <console type='pty'>
      <target type='serial' port='0'/>
    </console>
```

(I think the `<model name='isa-serial'>` is being added by libvirt later, and is the default.)

So yes, the 3003 changes seem quite promising.

I'll try to find time to upgrade both my salt master (easy), and a libvirt host (a bit more complicated as I have other local patches) to 3003 and verify this functionality works for me.  And if it does work (and I can drop my local patch to force serial console on those libvirt hosts) then I'll close this issue.

Ewen

**Comment by ewenmcneill** [comment]
> I'll try to find time to upgrade both my salt master (easy), and a libvirt host (a bit more complicated as I have other local patches) to 3003 and verify this functionality works for me. And if it does work (and I can drop my local patch to force serial console on those libvirt hosts) then I'll close this issue.

Unfortunately it appears that [`virt.keys` is entirely broken on 3003](https://github.com/saltstack/salt/issues/60275) (crashes in creating external pillar at the top of the method), which functionally means that `virt.*` is entirely broken on my system (as it relies on the `virt.keys` for authentication to libvirt, and without it seems like the `virt.*` methods start making weird decisions like trying to create VM disk images that already exist even though they're raw LVM volumes that already exist). 

So I've not been able to test this serial console syntax, other than to note that it does need to be:

```
 - serials:
    - type: pty
- consoles:
    - type: pty
```

as best I can tell, otherwise the YAML doesn't parse.

Ewen

**Comment by cbosdo** [comment]
> it seems like the `virt.*` methods start making weird decisions like trying to create VM disk images that already exist even though they're raw LVM volumes that already exist)

Could you be more precise on that one? I am curious to see if it's a bug or due to some missing pool refresh.

**Comment by ewenmcneill** [comment]
> Could you be more precise on that one? I am curious to see if it's a bug or due to some missing pool refresh.

@cbosdo, fortunately I still had the terminal window from the earlier runs open, so below is an example of the output I was getting on my `virt.running` runs when first trying 3003 (it happened for *every* VM defined for that virt/VM host, with basically the same output, as they're all from the same template/pillar data).  There's the relevant template of jinja populated states below, of which the disk relevant part is:

```
      - disks:
{%         for disk in vm['disks']|sort %}
          - name: {{disk}}
            source_file: {{vm['disks'][disk]['lv']}}
            format: raw
            model: virtio
            device: disk
{%         endfor %}
```

which indeed doesn't have any disk size on them.  Because they're logical volumes, and (a) they already exist, and (b) when they don't, they're pre-created by an earlier step anyway.  So I'd never given `virt.running` the disk size, just told them it was a `raw` disk, which I understood to mean it'd just use what was there (at least that's how it worked in 2019.x and 3002).

My *assumption* near the start of this multi-hour attempt to get 3003 working was that `virt.running` misbehaviour was happening because `virt.keys` was failing (see https://github.com/saltstack/salt/issues/60275), meaning that the "current state" recovery wasn't working, and it was disappearing down a path of "no idea what the current state is, I guess everything is new" and then trying to create the disks. But I didn't spend that long trying to debug the `virt.running` path after I realised it was most likely caused by the `virt.keys` step failing in my configuration.

However checking the `libvirtd` setup now, I see that it's actually configured with a TCP listener (not a TLS listener) on 127.0.0.1, which in theory means the keys aren't even required.  So maybe there is something else going on.  (And it looks like I maybe got half way through converting this VM host to TLS key authentication to libvirt, adding `virt.keys`, and then got distracted.)

FTR, the same state entries, apart from the (now commented out) serial/consoles section worked fine on 2019/3002.  So if there's a change in how the disk needs to be specified when it's a case of "the disk backing store will already exist, please don't touch it, just asume it exists", then that needs to be more obviously documented in the release notes/documentation.  The current disk definition documentation:

https://docs.saltproject.io/en/latest/ref/modules/all/salt.modules.virt.html#init-disk-def

mentions IO control bits as new in 3003, but nothing else about changing disk definitions to prevent attempts at creating the disk images as new for 3003.

Ewen

<details><summary>virt.running stack trace</summary>

```
          ID: debian_unstable
    Function: virt.running
      Result: False
     Comment: An exception occurred in this state: Traceback (most recent call last):
                File "/usr/lib/python3/dist-packages/salt/state.py", line 2172, in call
                  *cdata["args"], **cdata["kwargs"]
                File "/usr/lib/python3/dist-packages/salt/loader.py", line 1235, in __call__
                  return self.loader.run(run_func, *args, **kwargs)
                File "/usr/lib/python3/dist-packages/salt/loader.py", line 2268, in run
                  return self._last_context.run(self._run_as, _func_or_method, *args, **kwargs)
                File "/usr/lib/python3/dist-packages/contextvars/__init__.py", line 38, in run
                  return callable(*args, **kwargs)
                File "/usr/lib/python3/dist-packages/salt/loader.py", line 2283, in _run_as
                  return _func_or_method(*args, **kwargs)
                File "/usr/lib/python3/dist-packages/salt/loader.py", line 2316, in wrapper
                  return f(*args, **kwargs)
                File "/usr/lib/python3/dist-packages/salt/states/virt.py", line 1024, in running
                  host_devices=host_devices,
                File "/usr/lib/python3/dist-packages/salt/states/virt.py", line 669, in defined
                  host_devices=host_devices,
                File "/usr/lib/python3/dist-packages/salt/loader.py", line 1235, in __call__
                  return self.loader.run(run_func, *args, **kwargs)
                File "/usr/lib/python3/dist-packages/salt/loader.py", line 2268, in run
                  return self._last_context.run(self._run_as, _func_or_method, *args, **kwargs)
                File "/usr/lib/python3/dist-packages/contextvars/__init__.py", line 38, in run
                  return callable(*args, **kwargs)
                File "/usr/lib/python3/dist-packages/salt/loader.py", line 2283, in _run_as
                  return _func_or_method(*args, **kwargs)
                File "/usr/lib/python3/dist-packages/salt/modules/virt.py", line 4218, in update
                  _qemu_image_create(all_disks[idx])
                File "/usr/lib/python3/dist-packages/salt/modules/virt.py", line 1479, in _qemu_image_create
                  " disk size and/or disk image argument".format(disk["filename"])
              salt.exceptions.CommandExecutionError: Unable to create new disk debian_unstable_root, please specify disk size and/or disk image argument
     Started: 12:50:30.047737
    Duration: 170.328 ms
     Changes:   
```
</details>

<details><summary>virt.running state template</summary>

```
{{vm_name}}:
    virt.running:
      - cpu: {{vm.get('cpus', 1)}}
      - mem: {{vm['ram_mb']}}
      - vm_type: kvm
      - disk_profile: default
      - disks:
{%         for disk in vm['disks']|sort %}
          - name: {{disk}}
            source_file: {{vm['disks'][disk]['lv']}}
            format: raw
            model: virtio
            device: disk
{%         endfor %}
# TODO: nic_profile could be "sv" or "dmz" instead of default,
#       but I think we override enough that default should work.
#
      - nic_profile: default
      - install: False
      - interfaces:
{%         for nic in vm['net']|sort %}
          - name: {{nic}}
            type: bridge
            source: {{vm['net'][nic]['bridge']}}
            mac: {{vm['net'][nic]['mac_address']}}
            model: virtio
{%         endfor %}
      # Enable serial console, requires Salt 3003+
      # https://github.com/saltstack/salt/issues/53081#issuecomment-841904453
      # Salt 3003 doesn't currently work, in virt.keys :-(
      # See: https://github.com/saltstack/salt/issues/60275
      #- serials:
      #    - type: pty
      #- consoles:
      #    - type: pty
```
</details>

<details><summary>libvirtd config and example of logical volume already on disk</summary>

```
root@VM_HOST:/etc/libvirt# netstat -nlp | grep tcp.*libvirt
tcp        0      0 127.0.0.1:16509         0.0.0.0:*               LISTEN      3088/libvirtd       
root@VM_HOST:/etc/libvirt# grep -v "^#" libvirtd.conf | grep '[a-z]'
listen_tls = 0
listen_tcp = 1
listen_addr = "127.0.0.1"
unix_sock_group = "libvirt"
unix_sock_ro_perms = "0777"
unix_sock_rw_perms = "0770"
auth_unix_ro = "none"
auth_unix_rw = "none"
auth_tcp = "none"
root@VM_HOST:/etc/libvirt# 
```

```
root@VM_HOST:/etc/libvirt# lvdisplay /dev/ssd/debian_unstable_root
  --- Logical volume ---
  LV Path                /dev/ssd/debian_unstable_root
  LV Name                debian_unstable_root
  VG Name                ssd
  LV UUID                xsxa97-ZMgk-iAqn-T9fi-xFvH-nAWI-Wvtcl6
  LV Write Access        read/write
  LV Creation host, time naosr620, 2019-05-16 14:51:18 +1200
  LV Status              available
  # open                 2
  LV Size                20.00 GiB
  Current LE             5120
  Segments               1
  Allocation             inherit
  Read ahead sectors     auto
  - currently set to     256
  Block device           253:15
   
root@VM_HOST:/etc/libvirt# 
```
</details>

**Comment by ewenmcneill** [comment]
This "disk creation attempt in 3003" is potentially a *different* bug from the one in `virt.keys` and the one in this issue (console parameters); if so it possibly should be pulled out to a new issue. (I'm not doing that now, as I'm still unsure, and it's late here so I'm not debugging it in full today to make the call one way or another).

But I do note that:

https://github.com/saltstack/salt/blob/b88a23323783dd8ded7a20954f5da112fcc89b14/salt/modules/virt.py#L2894-L2898

and

https://github.com/saltstack/salt/blob/b88a23323783dd8ded7a20954f5da112fcc89b14/salt/modules/virt.py#L4213-L4218

use different tests for "does this thing already exist".  In particular `os.path.isfile` is *not* identical to `os.path.exists` when the relevant thing pointed at is a not actually a plain file.  And it looks like that second one is the path being traversed -- via modules/virt.py:4218 in the stack trace:

In particular in my case, they're symlinks to *block devices*, for which  I'd expect`os.path.isfile` to return false:

```
root@VM_HOST:/etc/libvirt# ls -ld /dev/ssd/debian_unstable_root
lrwxrwxrwx 1 root root 8 May 25 11:16 /dev/ssd/debian_unstable_root -> ../dm-15
root@VM_HOST:/etc/libvirt# ls -ld /dev/dm-15 
brw-rw---- 1 libvirt-qemu kvm 253, 15 May 31 20:22 /dev/dm-15
root@VM_HOST:/etc/libvirt# test -f /dev/ssd/debian_unstable_root || echo "Not a file"
Not a file
root@VM_HOST:/etc/libvirt# test -f /dev/dm-15 || echo "Not a file"
Not a file
root@VM_HOST:/etc/libvirt# 
```

So maybe this *is* a new regression separate from the rest?

Ewen

**Comment by cbosdo** [comment]
Ewen, that disk issue seems a different bug to me. I never tested using LVs and block devices using the `source_file` directly. As a workaround creating a storage pool of logical type and use volumes in it should work.

```yaml
ssd_pool:
  virt.pool_running:
    - name: ssd
    - ptype: logical
    - source:
        devices:
          - path: /dev/sdb
```

The disks would just be something like:

```yaml
      - disks:
{%         for disk in vm['disks']|sort %}
          - name: {{disk}}
            pool: {{vm['disks'][disk][vg]}}
            source_file: {{vm['disks'][disk]['lv']}}
            format: raw
            model: virtio
            device: disk
{%         endfor %}
```

with the pillar being changed to:

```yaml
vm:
    debian_unstable:
        domain: EXAMPLE.COM
        ram_mb: 2048
        net:
            eth0:
                ipv4_address: 172.20.A.B
                mac_address:  00:16:3e:aa:bb:cc
                bridge:       br-sv
        disks:
            vda:
                mountat: /
                vg:     ssd
                lv:      debian_unstable_root
                size:    20G
            vdb:
                mountat: swap
                vg:     ssd
                lv:      debian_unstable_swap
                size:    512M

```

I will try to reproduce your setup locally and see what goes west.

**Comment by cbosdo** [comment]
@ewenmcneill let's move the disk discussion to issue #60296. I managed to reproduce and your guess about `isfile()` is right: doesn't work with block devices and symlinks.

**Comment by ewenmcneill** [comment]
Closing issue (which I created) as (a) the `serials` / `consoles` feature was implemented in 3003 (see https://github.com/saltstack/salt/issues/53081#issuecomment-832459670), and (b) by 3004 I got salt working with this updated feature (removing my locally patched work around), after the disk image reference issue (#60296) was resolved.

So this request did get implemented, somewhere back in 2021.  Thanks!

Ewen
